src/server/worker/utils/utils.go

```diff
@@ -0,0 +1,662 @@
+package utils
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/user"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	"github.com/gogo/protobuf/types"
+	"gopkg.in/go-playground/webhooks.v5/github"
+	"gopkg.in/src-d/go-git.v4"
+	gitPlumbing "gopkg.in/src-d/go-git.v4/plumbing"
+
+	col "github.com/pachyderm/pachyderm/src/server/pkg/collection"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	"github.com/pachyderm/pachyderm/src/server/pkg/exec"
+	"github.com/pachyderm/pachyderm/src/server/pkg/hashtree"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsdb"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsutil"
+	filesync "github.com/pachyderm/pachyderm/src/server/pkg/sync"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+	"github.com/pachyderm/pachyderm/src/server/worker/common"
+	"github.com/pachyderm/pachyderm/src/server/worker/logs"
+	"github.com/pachyderm/pachyderm/src/server/worker/stats"
+)
+
+const (
+	// The maximum number of concurrent download/upload operations
+	concurrency = 100
+
+	chunkPrefix = "/chunk"
+	mergePrefix = "/merge"
+)
+
+type Utils interface {
```

Comment:
I'm not sure if there's a ton of point in this interface, I can't imagine us having a different implementation of it, and it's not needed for abstraction purposes, since the fields of the struct are already private. It's also pretty easy to add the interface after the fact.

I also think it's a little weird to have a struct / interface called `utils`, since I don't generally expect utils to be stateful, how would you feel about calling it `driver`? I think this is pretty similar to `pfs.Driver`, not that I love that name, but I think at this point it's established enough in the codebase and gets the concept across.

Reply:
I think this interface will primarily be useful for unit-testing other modules that depend on this.  Especially because several of the functions provided here are fairly heavyweight, I think this is an important ability.

I agree about the name and `driver` seems much better.

Reply:
Makes sense.